File: Dart/7-Trees/Problems/4-BinaryTreeLevelOrderTraversal/BinaryTreeLevelOrderTraversal.dart
// Dart program for Binary Tree Level Order Traversal using multiple approaches
// Real-world applications include hierarchical processing, scheduling tasks, and AI decision trees.

import 'dart:collection';

// Definition for a binary tree node
class TreeNode {
  int val; // Value of the node
  TreeNode? left; // Left child node
  TreeNode? right; // Right child node
  TreeNode(this.val, [this.left, this.right]); // Constructor to create a node
}

// 1. BFS Using Queue (Standard Level Order Traversal)
// This approach uses Breadth-First Search (BFS) with a queue to traverse the tree level by level.
// It processes nodes at each level before moving to the next level.
List<List<int>> levelOrderBFS(TreeNode? root) {
  List<List<int>> result = []; // Initialize the result list to store level-wise node values
  if (root == null) return result; // If the tree is empty, return an empty result

  Queue<TreeNode> queue = Queue<TreeNode>(); // Create a queue for BFS traversal
  queue.add(root); // Add the root node to the queue

  while (queue.isNotEmpty) {
    int levelSize = queue.length; // Get the number of nodes at the current level
    List<int> level = []; // Initialize a list to store node values at the current level

    // Process all nodes at the current level
    for (int i = 0; i < levelSize; i++) {
      TreeNode node = queue.removeFirst(); // Dequeue a node
      level.add(node.val); // Add the node's value to the current level list

      // Enqueue the left and right children if they exist
      if (node.left != null) queue.add(node.left!);
      if (node.right != null) queue.add(node.right!);
    }

    result.add(level); // Add the current level's node values to the result list
  }
  return result; // Return the level-wise node values
}

// 2. Recursive DFS Approach
// This approach uses Depth-First Search (DFS) recursively to traverse the tree.
// It maintains a level parameter to track the depth of each node and adds nodes to the corresponding level in the result list.
void dfs(TreeNode? node, int level, List<List<int>> result) {
  if (node == null) return; // Base case: if node is null, return

  // If the current level doesn't exist in the result list, create a new list for it
  if (result.length == level) result.add([]);

  result[level].add(node.val); // Add the node's value to the corresponding level list

  // Recursively traverse the left and right subtrees
  dfs(node.left, level + 1, result);
  dfs(node.right, level + 1, result);
}

List<List<int>> levelOrderDFS(TreeNode? root) {
  List<List<int>> result = []; // Initialize the result list
  dfs(root, 0, result); // Start the DFS traversal from the root at level 0
  return result; // Return the level-wise node values
}

// 3. Zigzag Level Order Traversal
// This approach is similar to BFS but alternates the order of node values at each level (left to right, right to left).
List<List<int>> zigzagLevelOrder(TreeNode? root) {
  List<List<int>> result = []; // Initialize the result list
  if (root == null) return result; // If the tree is empty, return an empty result

  Queue<TreeNode> queue = Queue<TreeNode>(); // Create a queue for BFS traversal
  queue.add(root); // Add the root node to the queue
  bool leftToRight = true; // Flag to track the direction of traversal (left to right or right to left)

  while (queue.isNotEmpty) {
    int levelSize = queue.length; // Get the number of nodes at the current level
    List<int> level = List.filled(levelSize, 0); // Initialize a list with fixed size to store node values at the current level

    // Process all nodes at the current level
    for (int i = 0; i < levelSize; i++) {
      TreeNode node = queue.removeFirst(); // Dequeue a node
      int index = leftToRight ? i : (levelSize - 1 - i); // Calculate the index based on the traversal direction
      level[index] = node.val; // Add the node's value to the level list at the calculated index

      // Enqueue the left and right children if they exist
      if (node.left != null) queue.add(node.left!);
      if (node.right != null) queue.add(node.right!);
    }

    result.add(level); // Add the current level's node values to the result list
    leftToRight = !leftToRight; // Toggle the traversal direction for the next level
  }
  return result; // Return the zigzag level-wise node values
}

// 4. Reverse Level Order Traversal
// This approach performs standard BFS and then reverses the result list to get the level order traversal in reverse.
List<List<int>> reverseLevelOrder(TreeNode? root) {
  List<List<int>> result = levelOrderBFS(root); // Perform standard BFS
  return result.reversed.toList(); // Reverse the result list and return it
}

// 5. Using a HashMap to Store Levels
// This approach uses a HashMap to store node values at each level.
// The keys of the HashMap represent the level, and the values are lists of node values at that level.
Map<int, List<int>> levelOrderWithMap(TreeNode? root) {
  Map<int, List<int>> levelMap = {}; // Initialize a HashMap to store level-wise node values

  // Recursive function to traverse the tree and store node values in the HashMap
  void traverse(TreeNode? node, int level) {
    if (node == null) return; // Base case: if node is null, return

    // If the current level doesn't exist in the HashMap, create a new list for it
    if (!levelMap.containsKey(level)) {
      levelMap[level] = [];
    }

    levelMap[level]!.add(node.val); // Add the node's value to the corresponding level list in the HashMap

    // Recursively traverse the left and right subtrees
    traverse(node.left, level + 1);
    traverse(node.right, level + 1);
  }

  traverse(root, 0); // Start the traversal from the root at level 0
  return levelMap; // Return the HashMap containing level-wise node values
}

void main() {
  TreeNode root = TreeNode(1,
      TreeNode(2, TreeNode(4), TreeNode(5)),
      TreeNode(3, null, TreeNode(6, null, TreeNode(7))));

  print("Level Order Traversal (BFS Queue): ${levelOrderBFS(root)}");
  print("Level Order Traversal (DFS Recursive): ${levelOrderDFS(root)}");
  print("Zigzag Level Order Traversal: ${zigzagLevelOrder(root)}");
  print("Reverse Level Order Traversal: ${reverseLevelOrder(root)}");
  print("Level Order with HashMap: ${levelOrderWithMap(root)}");
}

File: Python/7-Trees/Problems/4-BinaryTreeLevelOrderTraversal/BinaryTreeLevelOrderTraversal.py
from typing import List, Optional
from collections import deque
import json
import threading
import asyncio
import concurrent.futures

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        """
        Initializes a TreeNode object.

        Args:
            val (int): The value of the node. Defaults to 0.
            left (Optional[TreeNode]): The left child of the node. Defaults to None.
            right (Optional[TreeNode]): The right child of the node. Defaults to None.
        """
        self.val = val
        self.left = left
        self.right = right

# Approach 1: BFS Using Queue
def level_order_bfs(root: Optional[TreeNode]) -> List[List[int]]:
    """
    Performs a level order traversal of a binary tree using Breadth-First Search (BFS).
    It processes nodes level by level, adding the value of each node to a list
    representing that level.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        List[List[int]]: A list of lists, where each inner list represents the values
                         of the nodes at a given level.
    """
    if not root:
        return []  # Base case: if the root is None, return an empty list
    result = []  # Initialize the result list to store the level order traversal
    queue = deque([root])  # Initialize a queue with the root node

    while queue:
        level = []  # List to store nodes at the current level
        for _ in range(len(queue)):  # Iterate through all nodes at the current level
            node = queue.popleft()  # Dequeue a node
            level.append(node.val)  # Add the node's value to the current level list
            if node.left:
                queue.append(node.left)  # Enqueue the left child if it exists
            if node.right:
                queue.append(node.right)  # Enqueue the right child if it exists
        result.append(level)  # Add the list of nodes at the current level to the result
    return result  # Return the 2D list representing the level order traversal

# Approach 2: Recursive DFS
def level_order_dfs(root: Optional[TreeNode]) -> List[List[int]]:
    """
    Performs a level order traversal of a binary tree using recursive Depth-First Search (DFS).
    It visits nodes recursively and keeps track of the depth to store the node's value
    in the correct level list.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        List[List[int]]: A list of lists, where each inner list represents the values
                         of the nodes at a given level.
    """
    result = []  # Initialize the result list

    def dfs(node: Optional[TreeNode], depth: int):
        """
        Recursive helper function to perform DFS.

        Args:
            node (Optional[TreeNode]): The current node being visited.
            depth (int): The depth of the current node.
        """
        if not node:
            return  # Base case: if the node is None, return
        if len(result) <= depth:
            result.append([])  # If the level list doesn't exist, create it
        result[depth].append(node.val)  # Add the node's value to the corresponding level list
        dfs(node.left, depth + 1)  # Recursively visit the left child
        dfs(node.right, depth + 1)  # Recursively visit the right child

    dfs(root, 0)  # Start the DFS traversal from the root at depth 0
    return result  # Return the 2D list representing the level order traversal

# Approach 3: Multi-threaded BFS
def level_order_multithreaded(root: Optional[TreeNode]) -> List[List[int]]:
    """
    Attempts to perform a level order traversal of a binary tree using multiple threads.
    However, due to Python's Global Interpreter Lock (GIL), this approach does not
    achieve true parallelism and may not provide significant performance improvements.
    It divides the processing of each level into a separate thread.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        List[List[int]]: A list of lists, where each inner list represents the values
                         of the nodes at a given level.
    """
    if not root:
        return []  # Base case: if the root is None, return an empty list
    result = []  # Initialize the result list
    queue = deque([root])  # Initialize a queue with the root node

    def process_level():
        """
        Helper function to process a single level of the tree.
        """
        level = []
        for _ in range(len(queue)):
            node = queue.popleft()
            level.append(node.val)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        result.append(level)

    while queue:
        thread = threading.Thread(target=process_level)  # Create a thread for processing the current level
        thread.start()  # Start the thread
        thread.join()  # Wait for the thread to complete before moving to the next level
    return result  # Return the 2D list

# Approach 4: Async Level Order Traversal
async def level_order_async(root: Optional[TreeNode]) -> List[List[int]]:
    """
    Performs a level order traversal of a binary tree using asynchronous operations.
    This approach is similar to the multi-threaded version but uses `asyncio` instead
    of threads.  Like multithreading, it may not provide significant performance gains
    for CPU-bound tasks due to the GIL, but it's included for completeness and to
    demonstrate an alternative approach.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        List[List[int]]: A list of lists, where each inner list represents the values
                         of the nodes at a given level.
    """
    if not root:
        return []
    result = []
    queue = deque([root])

    async def process_level():
        level = []
        for _ in range(len(queue)):
            node = queue.popleft()
            level.append(node.val)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        result.append(level)

    while queue:
        await process_level()  # Await the completion of processing the current level
    return result

# Approach 5: JSON Serialization
def level_order_json(root: Optional[TreeNode]) -> List[List[int]]:
    """
    Performs a level order traversal of a binary tree by serializing the tree to a JSON
    string and then deserializing it back into a tree structure before performing a standard
    BFS traversal.  This approach is highly inefficient as it involves unnecessary
    serialization and deserialization.  It's included as a demonstration of an alternative
     (but not recommended) approach.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        List[List[int]]: A list of lists, where each inner list represents the values
                         of the nodes at a given level.
    """
    if not root:
        return []
    json_data = json.dumps(tree_to_dict(root))  # Serialize the tree to a JSON string
    root_dict = json.loads(json_data)  # Deserialize the JSON string back to a dictionary
    new_root = dict_to_tree(root_dict)  # Convert the dictionary to a TreeNode
    return level_order_bfs(new_root)  # Perform BFS on the reconstructed tree

# Helper functions for JSON serialization
def tree_to_dict(root: Optional[TreeNode]) -> dict:
    """
    Helper function to convert a binary tree to a dictionary representation.
    This is used for JSON serialization.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        dict: A dictionary representation of the tree.
    """
    if not root:
        return None
    return {"val": root.val, "left": tree_to_dict(root.left), "right": tree_to_dict(root.right)}

def dict_to_tree(data: Optional[dict]) -> Optional[TreeNode]:
    """
    Helper function to convert a dictionary representation of a tree back to a binary tree.
    This is used for JSON deserialization.

    Args:
        data (Optional[dict]): The dictionary representation of the tree.

    Returns:
        Optional[TreeNode]: The root of the reconstructed binary tree.
    """
    if not data:
        return None
    return TreeNode(data["val"], dict_to_tree(data["left"]), dict_to_tree(data["right"]))

# Sample Usage
tree = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))
print("BFS Level Order:", level_order_bfs(tree))
print("DFS Level Order:", level_order_dfs(tree))
print("Multi-threaded Level Order:", level_order_multithreaded(tree))
print("JSON Serialized Level Order:", level_order_json(tree))

# Run Async Level Order
async def run_async():
    result = await level_order_async(tree)
    print("Async Level Order:", result)

asyncio.run(run_async())
